File: Niveau 4/Niveau 4.2/carton.hs
--------------------------------------------------------------------------------
-- Name:        Retrouver un produit
-- Purpose:     France-IOI niveau 4.2
--
-- Author:      Sébastien
--
-- Created:     27/01/2023
--------------------------------------------------------------------------------

import Control.Monad
import Data.List
import Data.Array.Unboxed

main :: IO ()
main = do
    nbObjets <- readLn :: IO Int
    contenants <- fmap (listArray (1, nbObjets) . map read . words) getLine :: IO (Array Int Int)
    
    nbRecherches <- readLn :: IO Int
    replicateM_ nbRecherches
        (do
            objet1:objet2:_ <- fmap (map read . words) getLine :: IO [Int]
            let peres = zip (listerPeres contenants objet1) (listerPeres contenants objet2)
                communs = takeWhile (\(x,y) -> x == y) peres
            print $ snd (last communs)
        )
    

listerPeres :: Array Int Int -> Int -> [Int]
listerPeres _ 0 = [0]
listerPeres contenants n = (listerPeres contenants (contenants!n)) ++ [n]

File: Niveau 4/Niveau 4.2/carton.py
#-------------------------------------------------------------------------------
# Name:        Carton commun
# Purpose:     France-IOI niveau 4.2
#
# Author:      Sébastien
#
# Created:     26/01/2023
#-------------------------------------------------------------------------------

import time

def main():
    nbObjets = int(input())
    contenants = [0] + list(map(int, input().split()))

    nbRecherches = int(input())
    for n in range(nbRecherches):
        objet1, objet2 = map(int, input().split())

        trace = [False]*(nbObjets+1)
        noeud = objet1
        trace[noeud] = True
        while noeud > 0:
            noeud = contenants[noeud]
            trace[noeud] = True

        noeud = objet2
        while not trace[noeud]:
            noeud = contenants[noeud]

        print(noeud)

if __name__ == '__main__':
    start = time.time()
    main()
    end = time.time()
    print("Temps d'exécution :", end-start, "s")

"""
8
3 3 7 3 6 7 0 0
3
1 5
5 7
2 8
"""